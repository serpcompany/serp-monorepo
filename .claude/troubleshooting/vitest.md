https://vitest.dev/guide/common-errors.html