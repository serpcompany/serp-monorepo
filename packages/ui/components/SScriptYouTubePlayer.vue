<script setup lang="ts">
  import type { HTMLAttributes, ImgHTMLAttributes, Ref } from 'vue'
  import {
    useHead,
    useScriptTriggerElement,
    useScriptYouTubePlayer,
  } from '#imports'
  import { defu } from 'defu'
  /// <reference types="youtube" />
  import { computed, onMounted, ref } from 'vue'

  const props = withDefaults(
    defineProps<{
      placeholderAttrs?: ImgHTMLAttributes
      rootAttrs?: HTMLAttributes
      aboveTheFold?: boolean
      videoId: string
      playerVars?: YT.PlayerVars
      width?: number
      height?: number
    }>(),
    {
      // Use the manual trigger for explicit load control
      trigger: 'manual',
      playerVars: { autoplay: 0, playsinline: 1 },
      width: 640,
      height: 360,
      placeholderAttrs: () => ({}),
      rootAttrs: () => ({}),
    },
  )

  const emits = defineEmits<{
    'ready': [e: YT.PlayerEvent]
    'state-change': [e: YT.OnStateChangeEvent, target: YT.Player]
    'error': [e: YT.OnErrorEvent, target: YT.Player]
  }>()

  const events: (keyof YT.Events)[] = ['onReady', 'onStateChange', 'onError']

  const rootEl = ref<HTMLElement>()
  const youtubeEl = ref<HTMLElement>()
  const ready = ref(false)

  // Create a manual trigger promise
  const trigger = useScriptTriggerElement({
    trigger: 'manual',
    el: rootEl,
  })

  // Load & status handlers from composable
  const { onLoaded, status, load } = useScriptYouTubePlayer({
    scriptOptions: { trigger },
  })

  const player: Ref<YT.Player | undefined> = ref()

  onMounted(async () => {
    // Immediately inject the YouTube IFrame API script
    load()

    // Once loaded, hydrate the player
    onLoaded(async (instance) => {
      const YouTube =
        instance.YT instanceof Promise ? await instance.YT : instance.YT
      await new Promise<void>((resolve) => {
        if (typeof YT.Player === 'undefined')
          YouTube.ready(resolve)
        else resolve()
      })
      player.value = new YT.Player(youtubeEl.value!, {
        videoId: props.videoId,
        playerVars: props.playerVars,
        width: props.width,
        height: props.height,
        events: Object.fromEntries(
          events.map(event => [
            event,
            (e: unknown) => {
              const emitEventName = event
                .replace(/([A-Z])/g, '-$1')
                .replace('on-', '')
                .toLowerCase()

              // Map the event names to the correct emit types
              if (event === 'onReady') {
                emits('ready', e as YT.PlayerEvent)
                ready.value = true
              }
              else if (event === 'onStateChange') {
                emits('state-change', e as YT.OnStateChangeEvent, player.value!)
              }
              else if (event === 'onError') {
                emits('error', e as YT.OnErrorEvent, player.value!)
              }
            },
          ]),
        ),
      })
    })
  })

  defineExpose({ player })

  const rootAttrs = computed(() =>
    defu(props.rootAttrs, {
      'aria-busy': status.value === 'loading',
      'aria-label':
        status.value === 'awaitingLoad'
          ? 'YouTube Player - Placeholder'
          : status.value === 'loading'
            ? 'YouTube Player - Loading'
            : 'YouTube Player - Loaded',
      'role': 'application',
      'style': {
        position: 'relative',
        backgroundColor: 'black',
        maxWidth: '100%',
        width: `${props.width}px`,
        aspectRatio: `${props.width}/${props.height}`,
      },
      ...(trigger instanceof Promise ? trigger.ssrAttrs || {} : {}),
    } as HTMLAttributes),
  )

  const placeholder = computed(
    () => `https://i.ytimg.com/vi_webp/${props.videoId}/sddefault.webp`,
  )

  if (import.meta.server) {
    useHead({
      link: [
        {
          rel: props.aboveTheFold ? 'preconnect' : 'dns-prefetch',
          href: 'https://i.ytimg.com',
        },
        props.aboveTheFold
          ? { rel: 'preload', as: 'image', href: placeholder.value }
          : {},
      ],
    })
  }

  const placeholderAttrs = computed(() =>
    defu(props.placeholderAttrs, {
      src: placeholder.value,
      alt: '',
      loading: props.aboveTheFold ? 'eager' : 'lazy',
      style: { width: '100%', objectFit: 'cover', height: '100%' },
    } as ImgHTMLAttributes),
  )
</script>

<template>
  <ClientOnly fallback-tag="div">
    <template #fallback>
      <img v-bind="placeholderAttrs" alt="Video Placeholder" />
    </template>
    <div ref="rootEl" v-bind="rootAttrs">
      <div
        ref="youtubeEl"
        style="width: 100%; height: 100%; position: absolute; top: 0; left: 0"
      ></div>
      <slot v-if="!ready" name="placeholder" :placeholder="placeholder">
        <img v-bind="placeholderAttrs" />
      </slot>
      <slot v-if="status === 'loading'" name="loading">
        <ScriptLoadingIndicator />
      </slot>
      <slot v-else-if="status === 'error'" name="error"></slot>
      <slot></slot>
    </div>
  </ClientOnly>
</template>
