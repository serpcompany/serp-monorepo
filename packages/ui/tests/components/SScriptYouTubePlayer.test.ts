import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import { describe, expect, it } from 'vitest';
import { ref } from 'vue';
import SScriptYouTubePlayer from '../../components/SScriptYouTubePlayer.vue';
import ComponentRender from '../componentRender';

if (typeof globalThis.YT === 'undefined') {
    globalThis.YT = {
      Player: class {
        constructor(el: any, options: any) {}
        playVideo() {}
        loadVideoById(id: string) {}
      },
      ready: (callback: () => void) => callback()
    };
  }

mockNuxtImport('useHead', () => () => {});
mockNuxtImport('useScriptTriggerElement', () => () => ({ ssrAttrs: {} }));

let youtubePlayerStatus = 'awaitingLoad';

describe('SScriptYouTubePlayer', () => {
  const baseProps = { videoId: 'dQw4w9WgXcQ', width: 640, height: 360 };

  it.each([
    // Default state: no slot overrides, status will be 'awaitingLoad'
    ['default (awaitingLoad)', { props: { ...baseProps } }],
    // Custom placeholder slot (still 'awaitingLoad')
    ['with custom placeholder slot', { props: { ...baseProps }, slots: { placeholder: () => '<img alt="Custom Placeholder" />' } }],
    // Loading state: simulate status 'loading'
    ['with loading slot', { props: { ...baseProps }, slots: { loading: () => '<div>Loading...</div>' } }],
    // Awaiting load state: simulate status 'awaitingLoad' with a custom slot
    ['with awaitingLoad slot', { props: { ...baseProps }, slots: { awaitingLoad: () => '<div>Awaiting load</div>' } }],
    // Error state: simulate status 'error'
    ['with error slot', { props: { ...baseProps }, slots: { error: () => '<div>Error occurred</div>' } }],
  ])('renders %s correctly', async (desc: string, options: { props: any, slots?: any }) => {
    // Determine status based on the provided slots
    let status = 'awaitingLoad';
    if (options.slots?.loading) status = 'loading';
    if (options.slots?.error) status = 'error';

    youtubePlayerStatus = status;

    mockNuxtImport('useScriptYouTubePlayer', () => () => ({
        onLoaded: (cb: Function) => {
          if (youtubePlayerStatus !== 'awaitingLoad') {
            cb({ YT: { Player: class {} } });
          }
        },
        status: ref(youtubePlayerStatus)
      }));

    const html = await ComponentRender(`SScriptYouTubePlayer ${desc}`, options, SScriptYouTubePlayer);
    expect(html).toMatchSnapshot();
  });
});
