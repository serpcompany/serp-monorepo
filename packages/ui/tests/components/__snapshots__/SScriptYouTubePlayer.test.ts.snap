// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SScriptYouTubePlayer > renders default (awaitingLoad) correctly 1`] = `
<div
  aria-busy="false"
  aria-label="YouTube Player - Placeholder"
  role="application"
  style="
    position: relative;
    background-color: black;
    max-width: 100%;
    width: 640px;
    aspect-ratio: 640 / 360;
  "
>
  <div style="
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0px;
    left: 0px;
  "></div>
  <img
    alt=""
    loading="lazy"
    src="https://i.ytimg.com/vi_webp/dQw4w9WgXcQ/sddefault.webp"
    style="
      width: 100%;
      object-fit: cover;
      height: 100%;
    "
  />
  <!-- v-if -->
</div>
`;

exports[`SScriptYouTubePlayer > renders with awaitingLoad slot correctly 1`] = `
<div
  aria-busy="false"
  aria-label="YouTube Player - Placeholder"
  role="application"
  style="
    position: relative;
    background-color: black;
    max-width: 100%;
    width: 640px;
    aspect-ratio: 640 / 360;
  "
>
  <div style="
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0px;
    left: 0px;
  "></div>
  <img
    alt=""
    loading="lazy"
    src="https://i.ytimg.com/vi_webp/dQw4w9WgXcQ/sddefault.webp"
    style="
      width: 100%;
      object-fit: cover;
      height: 100%;
    "
  />
  <!-- v-if -->
</div>
`;

exports[`SScriptYouTubePlayer > renders with custom placeholder slot correctly 1`] = `
<div
  aria-busy="false"
  aria-label="YouTube Player - Placeholder"
  role="application"
  style="
    position: relative;
    background-color: black;
    max-width: 100%;
    width: 640px;
    aspect-ratio: 640 / 360;
  "
>
  <div style="
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0px;
    left: 0px;
  "></div>
  &lt;img alt="Custom Placeholder" /&gt;
  <!-- v-if -->
</div>
`;

exports[`SScriptYouTubePlayer > renders with error slot correctly 1`] = `
<div
  aria-busy="false"
  aria-label="YouTube Player - Loaded"
  role="application"
  style="
    position: relative;
    background-color: black;
    max-width: 100%;
    width: 640px;
    aspect-ratio: 640 / 360;
  "
>
  <div style="
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0px;
    left: 0px;
  "></div>
  <img
    alt=""
    loading="lazy"
    src="https://i.ytimg.com/vi_webp/dQw4w9WgXcQ/sddefault.webp"
    style="
      width: 100%;
      object-fit: cover;
      height: 100%;
    "
  />
  &lt;div&gt;Error occurred&lt;/div&gt;
</div>
`;

exports[`SScriptYouTubePlayer > renders with loading slot correctly 1`] = `
<div
  aria-busy="true"
  aria-label="YouTube Player - Loading"
  role="application"
  style="
    position: relative;
    background-color: black;
    max-width: 100%;
    width: 640px;
    aspect-ratio: 640 / 360;
  "
>
  <div style="
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0px;
    left: 0px;
  "></div>
  <img
    alt=""
    loading="lazy"
    src="https://i.ytimg.com/vi_webp/dQw4w9WgXcQ/sddefault.webp"
    style="
      width: 100%;
      object-fit: cover;
      height: 100%;
    "
  />
  &lt;div&gt;Loading...&lt;/div&gt;
</div>
`;

exports[`sScriptYouTubePlayer > renders default (awaitingLoad) correctly 1`] = `
<div
  aria-busy="false"
  aria-label="YouTube Player - Placeholder"
  role="application"
  style="
    position: relative;
    background-color: black;
    max-width: 100%;
    width: 640px;
    aspect-ratio: 640 / 360;
  "
>
  <div style="
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0px;
    left: 0px;
  "></div>
  <img
    alt=""
    loading="lazy"
    src="https://i.ytimg.com/vi_webp/dQw4w9WgXcQ/sddefault.webp"
    style="
      width: 100%;
      object-fit: cover;
      height: 100%;
    "
  />
  <!-- v-if -->
</div>
`;

exports[`sScriptYouTubePlayer > renders with awaitingLoad slot correctly 1`] = `
<div
  aria-busy="false"
  aria-label="YouTube Player - Placeholder"
  role="application"
  style="
    position: relative;
    background-color: black;
    max-width: 100%;
    width: 640px;
    aspect-ratio: 640 / 360;
  "
>
  <div style="
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0px;
    left: 0px;
  "></div>
  <img
    alt=""
    loading="lazy"
    src="https://i.ytimg.com/vi_webp/dQw4w9WgXcQ/sddefault.webp"
    style="
      width: 100%;
      object-fit: cover;
      height: 100%;
    "
  />
  <!-- v-if -->
</div>
`;

exports[`sScriptYouTubePlayer > renders with custom placeholder slot correctly 1`] = `
<div
  aria-busy="false"
  aria-label="YouTube Player - Placeholder"
  role="application"
  style="
    position: relative;
    background-color: black;
    max-width: 100%;
    width: 640px;
    aspect-ratio: 640 / 360;
  "
>
  <div style="
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0px;
    left: 0px;
  "></div>
  &lt;img alt="Custom Placeholder" /&gt;
  <!-- v-if -->
</div>
`;

exports[`sScriptYouTubePlayer > renders with error slot correctly 1`] = `
<div
  aria-busy="false"
  aria-label="YouTube Player - Loaded"
  role="application"
  style="
    position: relative;
    background-color: black;
    max-width: 100%;
    width: 640px;
    aspect-ratio: 640 / 360;
  "
>
  <div style="
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0px;
    left: 0px;
  "></div>
  <img
    alt=""
    loading="lazy"
    src="https://i.ytimg.com/vi_webp/dQw4w9WgXcQ/sddefault.webp"
    style="
      width: 100%;
      object-fit: cover;
      height: 100%;
    "
  />
  &lt;div&gt;Error occurred&lt;/div&gt;
</div>
`;

exports[`sScriptYouTubePlayer > renders with loading slot correctly 1`] = `
<div
  aria-busy="true"
  aria-label="YouTube Player - Loading"
  role="application"
  style="
    position: relative;
    background-color: black;
    max-width: 100%;
    width: 640px;
    aspect-ratio: 640 / 360;
  "
>
  <div style="
    width: 100%;
    height: 100%;
    position: absolute;
    top: 0px;
    left: 0px;
  "></div>
  <img
    alt=""
    loading="lazy"
    src="https://i.ytimg.com/vi_webp/dQw4w9WgXcQ/sddefault.webp"
    style="
      width: 100%;
      object-fit: cover;
      height: 100%;
    "
  />
  &lt;div&gt;Loading...&lt;/div&gt;
</div>
`;
